utils/stm-to-jitx.stanza

```diff
@@ -0,0 +1,373 @@
+    ;__________________________________
+    ;========= Pin Properties =========
+    ;‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾
+    ;Grab the pins defined for the pin properties.
+    val pin-properties = pin-properties(stm)
+    val generic-props = stm-to-jitx $ generic-pin(pin-properties)
+    val power-props = stm-to-jitx $ power-pin(pin-properties)
+    
+    ;Are the pads numeric (Double -> Int) or named (Ref -> String)?
+    val numeric-pads? = any?({pad(_) is Double}, rows(pin-properties))
+    val named-pads? = any?({pad(_) is String}, rows(pin-properties))
+    
+    ;If the pads are neither or both, then we are using malformed data.
+    if numeric-pads? and named-pads? :
+      fatal("Generated a micro-controller with both Int and Ref pads.") 
+    if not numeric-pads? and not named-pads? :
+      fatal("Generated a micro-controller without pads.") 
+    
+    ;Generate the pin-properties.
+    ;Cases are split between Int and Ref pads, and across generic and power properties usages.
+    if numeric-pads? :
+      pin-properties :
+        [pin:Ref | pads:Int ... | side:Dir | generic-pin:GenericPin | power-pin:PowerPin]
+        for row in rows(pin-properties) do :
+          val pin-ref = ref-ify(pin(row))
+          val pad-name = to-int(pad(row) as Double)
+          match(generic-props?(row), power-props?(row)) :
+            (g:True, p:True)   : [(pin-ref) | pad-name | side(row) | generic-props | power-props]
+            (g:True, p:False)  : [(pin-ref) | pad-name | side(row) | generic-props |      -     ]
+            (g:False, p:True)  : [(pin-ref) | pad-name | side(row) |       -       | power-props]
+            (g:False, p:False) : [(pin-ref) | pad-name | side(row) |       -       |      -     ]
+    else :
+      pin-properties :
+        [pin:Ref | pads:Ref ... | side:Dir | generic-pin:GenericPin | power-pin:PowerPin]
+        for row in rows(pin-properties) do :
+          val pin-ref = ref-ify(pin(row))
+          val pad-name = ref-ify(pad(row) as String)
+          match(generic-props?(row), power-props?(row)) :
+            (g:True, p:True)   : [(pin-ref) | pad-name | side(row) | generic-props | power-props]
+            (g:True, p:False)  : [(pin-ref) | pad-name | side(row) | generic-props |      -     ]
+            (g:False, p:True)  : [(pin-ref) | pad-name | side(row) |       -       | power-props]
+            (g:False, p:False) : [(pin-ref) | pad-name | side(row) |       -       |      -     ]
+    ;__________________________________
+    ;============ Supports ============
+    ;‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾
+    defn generate-mapped-ref (mapping:STMSupportMapping) -> JITXObject :
+      inside pcb-module :
+        require mygpio:gpio
+        match(require(mapping)) :
+          (req:String) :
+            val req-bundle = bundle-table[req]
+            require p0 : req-bundle
+            dot(p0, ref-ify(pin(mapping)))
+          (req:False) :
+            dot(self, ref-ify(pin(mapping)))
```

Comment:
The stanza code that does database to esir component mapping should be as general as possible to be reuse for other types of components with supports. json -> supports statements can be general.

`ref-ify` was a function to format refs as they appeared on the xml we scraped the data from for stms, we should use the default print of a ref to to represent a ref in json like "a.b[2].c.d[3]". I wrote a function to parse this back into a Ref

```
defn ref (text: String) -> Ref|False :
  if text == "false" :
    false
  else :
    ref-internal(text)

defn ref-internal (text: String) -> Ref :
  ;regex-match implementation incomplete so can do exact matches with "^mypattern$"
  ;val match-result = regex-match("^(\\w+).([\\w\.])+$", text)
  if text[length(text) - 1] == ']' :
    val idx = last(text, '[') as Int
    IndexRef(ref-internal(text[0 to idx]), to-int!(text[(idx + 1) to (length(text) - 1)]))
  else :
    val idx = last(text, '.')
    match(idx: Int) :
      FieldRef(ref-internal(text[0 to idx]), Ref(text[(idx + 1) to length(text)]))
    else :
      Ref(text)
```